app/controllers/declaration/FiscalInformationController.scala

```diff
@@ -43,10 +44,18 @@ class FiscalInformationController @Inject()(
     extends FrontendController(mcc) with I18nSupport with ModelCacheable {
 
   def displayPage(mode: Mode, itemId: String, fastForward: Boolean): Action[AnyContent] = (authenticate andThen journeyType) { implicit request =>
-    val cacheContainsFiscalReferenceData = request.cacheModel.itemBy(itemId).exists(_.additionalFiscalReferencesData.exists(_.references.nonEmpty))
+    def cacheContainsFiscalReferenceData = request.cacheModel.itemBy(itemId).exists(_.additionalFiscalReferencesData.exists(_.references.nonEmpty))
+    def cacheItemIneligibleForOSR =
+      request.cacheModel
+        .itemBy(itemId)
+        .flatMap(_.procedureCodes)
+        .flatMap(_.procedureCode)
+        .exists(code => !ProcedureCodesData.osrProcedureCodes.contains(code))
 
     if (fastForward && cacheContainsFiscalReferenceData) {
       navigator.continueTo(routes.AdditionalFiscalReferencesController.displayPage(mode, itemId))
+    } else if (fastForward && cacheItemIneligibleForOSR) {
+      navigator.continueTo(routes.ProcedureCodesController.displayPage(mode, itemId))
```

Comment:
Just to confirm this redirection is correct? I thought that ProcedureCodes is a page before Fiscal Information

Reply:
That is for "back link" navigation.   Perhaps the "fastforward" variable could be better named "back navigating"?

The logic is when coming back from the next page (Commodity Code) we direct to FiscalInfo.   We then either redirect one page forward to the VAT page if there is existing fiscal data (old implementation), re-direct to the page before if OSR not valid due to existing Procedure Code (new logic) or stay here.

I've tested this manually and it all appears to work.   It is a bit "clunky" though.

Reply:
There are no existing tests in the FiscalControllerSpec for "fast-forward" BTW - adding those now including the new ProcedureCode invalid logic.

Reply:
Ahh, I forgot about it. Thanks for explanation!

That's right.